### DISCUSSION TITLE
IOCP with multiple threads.

### DISCUSSION THREAD

**Discussion by RCECoder** [libuv/libuv#4540] [General]
Any example showing how to use libuv with multiple threads on Windows? 
I see that `uv_default_loop ` calls `CreateIoCompletionPort ` with `NumberOfConcurrentThreads ` set to 1. And there is no option to set it.

**Comment by saghul** [comment]
Each libuv loop is single threaded, you cannot change that.

You can, however, run separate loops in separate threads.

**Reply by RCECoder** [reply]
> You can, however, run separate loops in separate threads.

This means each thread has its own completion port?
How would you communicate with the thread pool?

**Reply by saghul** [reply]
Yes, each loop has its own completion port. For communicating with the loop from an outside thread you can use uv_async_t. The internal thread pool already takes care of synchronization, you can schedule work there from the loop thread with uv_queue_work.

**Reply by RCECoder** [reply]
> Yes, each loop has its own completion port. For communicating with the loop from an outside thread you can use uv_async_t. The internal thread pool already takes care of synchronization, you can schedule work there from the loop thread with uv_queue_work.

Thanks for reply
Do you mean to use `uv_async_send` to wake up the loop then execute the callback?
But this function acts as "many to one" as per documentation. How can we make it executes the callback for each call or this is not possible to do?

**Reply by saghul** [reply]
What exactly are you trying to accomplish?

**Reply by RCECoder** [reply]
> What exactly are you trying to accomplish?

I want my server to be multi threaded.

**Reply by saghul** [reply]
One thing you can do is have a single thread to accept connections, and then use pipes to send the connections over to other worker threads using uv_write2 so they can handle the connection.

**Reply by RCECoder** [reply]
I haven't thought of this idea.
Any reason why a thread loop was designed as single-threaded?

**Reply by saghul** [reply]
libuv started out wrapping libev which was single threaded. It was designed to be the platform layer of NodeJS, which is single threaded (for the most part). The added complexity and synchronization necessary to make it work with multiple threads out of the box would likely be very taxing.

**Reply by bnoordhuis** [reply]
The observation at the time was that boost.asio pays a fairly hefty price in complexity and performance to support multi-threading even though almost everyone uses it in single-threaded mode.
